analysis/data_utils.py

```diff
@@ -311,3 +315,87 @@ def experiment_level_ranking(experiment_snapshots_df,
     pivot_table = experiment_pivot_table(experiment_snapshots_df,
                                          benchmark_level_ranking_function)
     return experiment_level_ranking_function(pivot_table)
+
+
+_DEFAULT_RARE_REGION_THRESHOLD = 0
+
+
+def get_rare_region_dict(benchmark_coverage_dict,
+                         threshold=_DEFAULT_RARE_REGION_THRESHOLD):
+    """Returns a dictionary containing the covering fuzzers for each
+    rare region, where the |threshold| defines which regions are rare."""
+    region_dict = defaultdict(list)
+    rare_region_dict = {}
+    threshold_count = math.ceil(len(benchmark_coverage_dict) * threshold)
+    if threshold_count == 0:
+        threshold_count = 1
+    for fuzzer in benchmark_coverage_dict:
+        for region in benchmark_coverage_dict[fuzzer]:
+            region_dict[region].append(fuzzer)
+    for region, fuzzers in region_dict.items():
+        if len(fuzzers) <= threshold_count:
+            rare_region_dict[region] = fuzzers
+    return rare_region_dict
+
+
+def get_rare_region_cov_df(rare_region_dict, fuzzer_names):
+    """Returns a DataFrame where the two columns are fuzzers and the number
+    of rare regions covered."""
+    fuzzers = {fuzzer_name: 0 for fuzzer_name in fuzzer_names}
+    for region in rare_region_dict:
+        for fuzzer in rare_region_dict[region]:
+            fuzzers[fuzzer] += 1
+    dict_to_transform = {'fuzzer': [], 'rare_region_covered': []}
+    for fuzzer, covered_num in fuzzers.items():
+        dict_to_transform['fuzzer'].append(fuzzer)
+        dict_to_transform['rare_region_covered'].append(covered_num)
+    return pd.DataFrame(dict_to_transform)
+
+
+def get_benchmark_cov_dict(coverage_dict, benchmark):
+    """Returns a dictionary to store the covered regions of each fuzzer.
+    Uses a set of tuples to store the covered regions."""
+    benchmark_cov_dict = {}
+    for key_pair, covered_regions in coverage_dict.items():
+        current_fuzzer, current_benchmark = key_pair.split()
+        if current_benchmark == benchmark:
+            covered_regions_in_set = set()
+            for region in covered_regions:
+                covered_regions_in_set.add(tuple(region))
+            benchmark_cov_dict[current_fuzzer] = covered_regions_in_set
+    return benchmark_cov_dict
+
+
+def get_correlation_table(benchmark_coverage_dict):
+    """Returns a table that shows the correlation between each two fuzzers.
+
+    The correlation table is a square matrix where each row and column
```

Comment:
It'd be interesting to see real examples of this, but based on the description ("each cell contains a number showing the regions covered by the fuzzer of the column but not by the fuzzer of the row") it feels like the reader needs to keep in mind which is the "row" fuzzer and the "column" fuzzer.

I wonder if it'd make sense to show in each cell what percentage of the two fuzzer's coverage is the same. While this'd be the "inverse" of the same info, it'd probably be easier to read due to the symmetry (doesn't matter what's in the row and in column). Also, that's what i'd associate more with something called a "correlation plot". wdty?

Reply:
That's true! And that's what I was trying to do in the first place. 
However, the goal of this matrix is to help the user determine, if they have used fuzzer-1, which fuzzer else should they use if they want to complements a combined result. 
Under such guideline, say we have `fuzzer1` and `fuzzer2`.
`fuzzer1` covers `region1`, `region2` and `region3`
`fuzzer2` covers `region2`
The question is that what are you going to put in the two cells for `fuzzer1` and `fuzzer2`? Is it going to be 100% or 33.3%? It depends on what is the main fuzzer you are using and what the additional fuzzer you want to complements. This is why I don't think the matrix should be symmetric.

Reply:
Yes, that's a good point!

> The question is that what are you going to put in the two cells for fuzzer1 and fuzzer2? Is it going to be 100% or 33.3%?

Do you mean according to the symmetric correlation suggestion? I'd put 33.33%, as in Cov(f1)∩Cov(f2) / Cov(f1)UCov(f2).

I'm not against the non-symmetric approach, I agree it captures more information. I'm just a bit worried that because it's  more difficult to read/interpret, people won't look at it much. Perhaps it's worth thinking about what are the most usual questions people ask that these plots can help with. E.g., one question could be: I see f1 and f2 did very well on this benchmark, I wonder if they discovered more or less the same parts of the code or totally different ones. For this, the symmetric correlation plot gives a good answer. But there might be other typical questions that it doesn't. Wdyt?

Reply:
Yeah I agree that it is harder to interpret if not symmetric. But I think the information of symmetric approach may be misleading sometimes.
e.g.
`f1` covers `region1` to `region99`
`f2` covers `region1` to `region200`
`f3` covers `region100` and `region99`
For correlation of `f1-f2`, Cov(f1)∩Cov(f2) / Cov(f1)UCov(f2) = 49.5%
For correlation of `f1-f3`, Cov(f1)∩Cov(f3) / Cov(f1)UCov(f3) = 1%
However, people won't want to use `f3` to complement `f1` since it can only measure 1 additional region. That's why I didn't use percentage but absolute value.
Wdyt?

Reply:
Yes, the symmetric thing alone wouldn't directly answer that question. Sgtm to go with the current implementation. Consider finding a more descriptive name for it though, because "correlation" table/plot is a thing, but this plot has not much to do with that if i understand correctly.

Also the description under the plot says "showing the percentage", but you've just mentioned "didn't use percentage but absolute value".

Reply:
Yeah I'll change it right away.